rust/otel-arrow-rust/src/otap/filter.rs

```diff
@@ -0,0 +1,60 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+use arrow::array::BooleanArray;
+use serde::Deserialize;
+
+pub mod logs;
+
+/// MatchType describes how we should match the String values provided
+#[derive(Debug, Clone, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum MatchType {
+    /// match on the string values exactly how they are defined
+    Strict,
+    /// apply string values as a regexp
+    Regexp,
+}
+
+/// enum that allows a field to have any type
+#[derive(Debug, Clone, PartialEq, Deserialize)]
+#[serde(untagged)]
+pub enum AnyValue {
+    /// string type
+    String(String),
+    /// int type
+    Int(i64),
+    /// double type
+    Double(f64),
+    /// boolean type
+    Boolean(bool),
+    /// array of any type
+    Array(Vec<AnyValue>),
+    /// keyvalue type
+    KeyValue(Vec<KeyValue>),
+}
+
+/// struct that represents attributes and other key value pairs
+#[derive(Debug, Clone, PartialEq, Deserialize)]
+pub struct KeyValue {
+    key: String,
+    value: AnyValue,
+}
+
+impl KeyValue {
+    /// create a new key value pair
+    #[must_use]
+    pub fn new(key: String, value: AnyValue) -> Self {
+        Self { key, value }
+    }
+}
+
+/// Finds all nulls and converts them to false values
+/// null values affect the filter computation as when
+/// we perform the not operation nothing happens to
+/// the null values
+#[must_use]
+pub fn nulls_to_false(a: &BooleanArray) -> BooleanArray {
+    let valid = arrow::compute::is_not_null(a).expect("is_not_null doesn't error"); // BooleanArray with no nulls
+    arrow::compute::and_kleene(a, &valid).expect("can combine two columns with equal length") // nulls become false; trues stay true
```

Comment:
Can these two "expect" statements be triggered by a misconfiguration of the `FilterProcessor`?
If so, we should replace them with one or two error returns.
We should only use `expect` in critical cases where the error indicates an internal logic problem.